storage/innobase/row/row0mysql.cc

```diff
@@ -2826,6 +2842,170 @@ row_add_table_to_background_drop_list(table_id_t table_id)
+		unlink(trun->file_name);
+	}
+	else {
+		truncate(trun->file_name, st.st_size - srv_async_truncate_size*1024*1024);
+		return true;
+	}
+
+already_dropped:
+	mutex_enter(&row_truncate_list_mutex);
+	UT_LIST_REMOVE(row_mysql_truncate_list, trun);
+	mutex_exit(&row_truncate_list_mutex);
+
+	ut_free(trun->file_name);
+	ut_free(trun);
+
+	return true;
+}
+
+/** The master thread in srv0srv.cc calls this when shutdown innodb, we delete all
+temp files left by unlink now, instead of truncate
+@return false if all files are deleted */
+bool
+row_truncate_file_for_mysql_in_background_shutdown(void)
+{
+	if (row_mysql_truncate_list_inited == FALSE)
+		return false;
+
+	row_mysql_truncate_t *truncate, *next;
+
+	mutex_enter(&row_truncate_list_mutex);
+
+	/* Look if the file already is in the truncate list */
+	for (truncate = UT_LIST_GET_FIRST(row_mysql_truncate_list);
+	     truncate != NULL;) {
+		next = UT_LIST_GET_NEXT(row_mysql_truncate_list, truncate);
+		UT_LIST_REMOVE(row_mysql_truncate_list, truncate);
+		unlink(truncate->file_name);
+		ut_free(truncate->file_name);
+		ut_free(truncate);
+		truncate = next;
+	}
+
+	mutex_exit(&row_truncate_list_mutex);
+	ut_a(UT_LIST_GET_LEN(row_mysql_truncate_list) == 0);
+	return false;
+}
+
+/** If a file is not yet in the truncate list, adds the file to the list of files
+which the master thread truncates in background.
+@return TRUE if the file was not yet in the list, and was added there */
+bool
+row_add_file_to_background_truncate_list(
+/*==================================*/
+	const char*	name)	/*!< in: file name */
+{
+	row_mysql_truncate_t* truncate;
+	ut_a(row_mysql_truncate_list_inited);
+
+	mutex_enter(&row_truncate_list_mutex);
+
+	/* Look if the file already is in the truncate list */
+	for (truncate = UT_LIST_GET_FIRST(row_mysql_truncate_list);
+	     truncate != NULL;
+	     truncate = UT_LIST_GET_NEXT(row_mysql_truncate_list, truncate)) {
+		if (strcmp(truncate->file_name, name) == 0) {
+			/* Already in the list */
+			mutex_exit(&row_drop_list_mutex);
+			return false;
+		}
+	}
+
+	truncate = static_cast<row_mysql_truncate_t*>(
+		ut_malloc_nokey(sizeof(row_mysql_truncate_t)));
+
+	truncate->file_name = mem_strdup(name);
+	truncate->fail_count = 0;
+
+	UT_LIST_ADD_LAST(row_mysql_truncate_list, truncate);
+
+	mutex_exit(&row_truncate_list_mutex);
+
+	return true;
+}
+
+/** add files under specific directory to the list which master thread would
+truncate in background. This is called in InnoDB bootstrap, to clean up orphan
+files left of last shutdown, so lock protection and duplicate check is not needed. */
+static void
+init_mysql_truncate_list(char* dir_path)
+{
+	row_mysql_truncate_t* truncate;
+	os_file_dir_t dir;
+	os_file_stat_t fileinfo;
+	dberr_t err = DB_SUCCESS;
+	ut_a(row_mysql_truncate_list_inited == FALSE);
+
+	dir = os_file_opendir(dir_path, true);
+	if (dir == NULL)
+		return;
+
+	while (fil_file_readdir_next_file(&err, dir_path, dir, &fileinfo) == 0) {
```

Comment:
I think that moving files to another directory only works if both directories are in the same file system.
I rather avoid this, and simply rely on the delete-on-close semantics.
`unlink()` should be fast if file handles remain open, and in my suggested approach, we would keep them open until we have finished truncating the file piece by piece.

Reply:
> I think that moving files to another directory only works if both directories are in the same file system.

   I think you are right , but if we do this in our product environment, people' space won't be truncated at first, and they can't write data or create new table due to lack of space. so we introduce a parameter to control this behavior. how do you think of that ?

Reply:
Thank you very much for reply. I think 'delete-on-close' is a better idea than the patch I did before and I will revise that part according to your suggestions. 
   But I have a question  after my study of dropping table in Mariadb. that is,
 
Not all dropped tables will be added to the row_mysql_drop_list, tables that are not added to the queue will also have the same problems before, the stack is:

	```
	#0  os_file_delete_func (name=0x7ff6200368b0 "./test/t1.ibd") at ../storage/innobase/os/os0file.cc:3173
	#1  0x00007ff67d52a650 in pfs_os_file_delete_func (key=47, name=0x7ff6200368b0 "./test/t1.ibd", src_file=0x7ff67da73da0 "../storage/innobase/fil/fil0fil.cc", src_line=2728)
	    at ..//storage/innobase/include/os0file.ic:505
	#2  0x00007ff67d532834 in fil_delete_tablespace (id=12, drop_ahi=false) at ..//storage/innobase/fil/fil0fil.cc:2728
	#3  0x00007ff67d36f3d6 in row_drop_table_for_mysql (name=0x7ff671c37420 "test/t1", trx=0x7ff6727321c8, sqlcom=SQLCOM_DROP_TABLE, create_failed=false, nonatomic=true) at ../storage/innobase/row/row0mysql.cc:3932
	#4  0x00007ff67d23016a in ha_innobase::delete_table (this=0x7ff620017860, name=0x7ff671c39040 "./test/t1", sqlcom=SQLCOM_DROP_TABLE) at ../storage/innobase/handler/ha_innodb.cc:12916
	#5  0x00007ff67d21cc67 in ha_innobase::delete_table (this=0x7ff620017860, name=0x7ff671c39040 "./test/t1") at ../storage/innobase/handler/ha_innodb.cc:13045
	#6  0x00007ff67cee84f4 in handler::ha_delete_table (this=0x7ff620017860, name=0x7ff671c39040 "./test/t1") at ../sql/handler.cc:4560
	#7  0x00007ff67cee2551 in ha_delete_table (thd=0x7ff620000b00, table_type=0x7ff67f828b00, path=0x7ff671c39040 "./test/t1", db=0x7ff671c38cd0, alias=0x7ff620017008, generate_warning=true) at ../sql/handler.cc:2529
	#8  0x00007ff67cca9fa5 in mysql_rm_table_no_locks (thd=0x7ff620000b00, tables=0x7ff620016fe0, if_exists=false, drop_temporary=false, drop_view=false, drop_sequence=false, dont_log_query=false, dont_free_locks=false)
	    at ../sql/sql_table.cc:2510
	#9  0x00007ff67cca8faf in mysql_rm_table (thd=0x7ff620000b00, tables=0x7ff620016fe0, if_exists=false, drop_temporary=false, drop_sequence=false) at ../sql/sql_table.cc:2124
	#10 0x00007ff67cbe3598 in mysql_execute_command (thd=0x7ff620000b00) at ..//sql/sql_parse.cc:5052
	#11 0x00007ff67cbecdaa in mysql_parse (thd=0x7ff620000b00, rawbuf=0x7ff620016f18 "drop table t1", length=13, parser_state=0x7ff671c3a4f0, is_com_multi=false, is_next_command=false) at ../sql/sql_parse.cc:8091
	#12 0x00007ff67cbd9fa2 in dispatch_command (command=COM_QUERY, thd=0x7ff620000b00, packet=0x7ff62000b1e1 "drop table t1", packet_length=13, is_com_multi=false, is_next_command=false) at ../sql/sql_parse.cc:1850
	#13 0x00007ff67cbd89d3 in do_command (thd=0x7ff620000b00) at ../sql/sql_parse.cc:1395
	#14 0x00007ff67cd41514 in do_handle_one_connection (connect=0x7ff67fc467d0) at ../sql/sql_connect.cc:1402
	#15 0x00007ff67cd41298 in handle_one_connection (arg=0x7ff67fc467d0) at ../sql/sql_connect.cc:1308
	#16 0x00007ff67d1f1d48 in pfs_spawn_thread (arg=0x7ff67fc4ed70) at ../storage/perfschema/pfs.cc:1862
	#17 0x00007ff67c0ece25 in start_thread () from /lib64/libpthread.so.0
	#18 0x00007ff67a34435d in clone () from /lib64/libc.so.6
```
how do you think of that ?

Reply:
I think separate dropping tables and truncating them background is more clearly in logic,  So can we add a thread to do truncate logic ?